Add Global Styles support for the Add To Cart Button block #5666

gigitux opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

This PR adds support for global style for the `Add To Cart Button` block.

Now, the user can edit the style for:
- text-color
- background-color
- border-radius
- font-size


<!-- Reference any related issues or PRs here -->
#4965 

### Screenshots

![image](https://user-images.githubusercontent.com/4463174/153617092-faebfb9f-6941-46ca-b871-e569de11a2b6.png)
*without global style*

![image](https://user-images.githubusercontent.com/4463174/153616344-f278a557-49d7-4968-838f-79e639849b9b.png)
*with global style*

### Testing

### Manual Testing

1. Be sure that you are on `WordPress 5.9`.
2. Install and enable the `Twenty Twenty-Two` theme.
3. Add the `All Product` block (this block contains `Add To Cart Button Block`) to a post.
4. Get the focus on `Add To Cart Button`.
5. On the right sidebar, personalize the styles of the block.
6. Go on the page and check if there are changes.
7. Reset to default using the `Reset` button from the different sections. 
8. Go to Dashboard and select Appearance > Editor (beta). At the top of the screen, select Home > Browser all templates > Single Post. When the page is loaded, add the block to the page.
9. On the Editor page click on the `Styles` icon on the right-top corner.
10. Verify that the `Add To Cart Button` is shown under the `Blocks` section. Personalize again the block.
11. Save your changes.
12. Go on the page created earlier and check if all styles are applied correctly.
13. Edit your previous post/page again.
14. Change again the styles.
15. Save your changes.
16. Check if these styles have priority over the styles from the Site Editor.


### Changelog

> Add Global Styles support to the Add To Cart Button block.

gigitux committed:
Add Global Styles support for the Add To Cart Button block #5666

Add Global Styles support for the Add To Cart Button block

gigitux committed:
Merge branch 'trunk' into add/5666-add-to-cart-button-2

dinhtungdu commented on assets/js/atomic/blocks/product-elements/button/block.js:
```diff
@@ -116,14 +151,22 @@ const AddToCartButton = ( { product } ) => {
 		<ButtonTag
 			aria-label={ buttonAriaLabel }
 			className={ classnames(
+				'wc-block-components-product-button__button--global-style',
```
Why do we need another class here? Scanning this PR, I see it's for Global Style. Is there any issue with the existing class (`wc-block-components-product-button__button`)?

gigitux replied:
I was convinced that this resolved the problem about the font size :(

I did another cycle of tests and I noticed that this is not true :/ 
I removed this new class and thanks for bringing up this!

dinhtungdu reviewed (commented):
> The problem is that this block uses padding that comes from the wp-button-link class: if we override padding property we lose the default padding value. In this case, the style of the button would be not great.

Can you explain this further? I tested adding padding to the block and didn't see issues related to `.wp-button-link` style.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/5666-add-to-cart-button-2

gigitux commented:
> > The problem is that this block uses padding that comes from the wp-button-link class: if we override padding property we lose the default padding value. In this case, the style of the button would be not great.
> 
> Can you explain this further? I tested adding padding to the block and didn't see issues related to `.wp-button-link` style.

Did you try with the `Global Styles`? 
![image](https://user-images.githubusercontent.com/4463174/153912776-e7c631ff-ce02-4746-ac0a-e5be0f104251.png)
 
With the current implementation, the current style will have always greater priority respect than the Global Style settings.
 
 If I remove the padding, the user can see the padding that inserts, but at this point, the default padding will be 0 and the button doesn't look great.

I wasn't able to find a good workaround, but maybe you have another idea to fix this
 
 
 
 Unfortunately, I noticed that we have the same problem with the font size :'(

gigitux committed:
remove support to font size

gigitux committed:
Merge branch 'add/5666-add-to-cart-button-2' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/5666-add-to-cart-button-2

dinhtungdu reviewed (commented):
> With the current implementation, the current style will have always greater priority respect than the Global Style settings.

@gigitux https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5843 is my workaround to fix the style priority, can you please take a look?

dinhtungdu committed:
try: fix the style override issue for add to cart button (#5843)

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/5666-add-to-cart-button-2

gigitux committed:
Merge branch 'add/5666-add-to-cart-button-2' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/5666-add-to-cart-button-2

gigitux commented:
> > With the current implementation, the current style will have always greater priority respect than the Global Style settings.
> 
> @gigitux #5843 is my workaround to fix the style priority, can you please take a look?

Thanks for your help! 🙏

dinhtungdu commented on assets/js/atomic/blocks/product-elements/button/supports.ts:
```diff
@@ -11,15 +11,10 @@ export const supports = {
-		typography: {
-			fontSize: true,
-			__experimentalSkipSerialization: true,
-		},
```
@gigitux Should we add the typo support back?

gigitux committed:
restore typography support

dinhtungdu reviewed (commented):
The font-weight of the Button is different between the Blocks Editor and the front end. Otherwise, this is LGTM

<img width="2008" alt="image" src="https://user-images.githubusercontent.com/5423135/154661343-1e0ebed9-476b-4f8b-9490-a5f2116bef5f.png">

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/5666-add-to-cart-button-2

gigitux committed:
fix font weight

gigitux commented:
Fixed with https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5816/commits/abcd96a31ee72c618a8379344e5395f4dfdca16d!

![image](https://user-images.githubusercontent.com/4463174/155149128-756b1115-c11f-48f5-8f7e-9fd86d5878a9.png)
*frontend side*

![image](https://user-images.githubusercontent.com/4463174/155149040-6ec102f7-7a3e-4242-ba39-5e81c2ebdfa7.png)
*editor side*

dinhtungdu reviewed (commented):
@gigitux Because we added font-size support, what do you think about adding font-weight support as well? Using 400 as the font weight for the add to cart button is a valid use case to me.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/5666-add-to-cart-button-2

gigitux committed:
add support for font weight

gigitux commented:
> @gigitux Because we added font-size support, what do you think about adding font-weight support as well? Using 400 as the font weight for the add to cart button is a valid use case to me.

I added the font-weight support. 
I added [this line](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5816/files#diff-95a248c54e90940a2c3a4595a96d039f8a70d8fb593258000c17c3e658cd2ea5R10) because I noticed that the button element used the Arial font.

## Before my changes 

![image](https://user-images.githubusercontent.com/4463174/155513161-6ca3d64f-f257-42ed-902a-a1cfef2da7ca.png)
*button element (third button)*
![image](https://user-images.githubusercontent.com/4463174/155513181-12386ce1-f743-477f-bccb-511fd7910028.png)
*anchor element (two first buttons)*

dinhtungdu commented on assets/js/atomic/blocks/product-elements/button/supports.ts:
```diff
@@ -0,0 +1,30 @@
+			__experimentalFontWeight: true,
+			__experimentalSkipSerialization: true,
+		},
+		__experimentalSelector:
+			'.wp-block-button.wc-block-components-product-button .wc-block-components-product-button__button',
```
Just FYI, on the server side, Global Style only cares about `__experimentalSelector` to generate CSS.

dinhtungdu reviewed (approved):
LGTM 🎉  :shipit:
